### PR TITLE
METRON-2217 Migrate current HBase client from HTableInterface to Table

### PR DESCRIPTION
**Pull request by mmiklavc**
## Contributor Comments

https://issues.apache.org/jira/browse/METRON-2217

**UPDATE** - Test plan has been added below. I ran through this plan and variations of it many times over as I went through testing.

I want to get this in front of people to start reviewing asap. It's going to take me a couple days to work through a reasonable test plan for this, but this should not hold up reviewing the approach. Of note is that the change from HTableInterface to Table by HBase has now shifted the burden of connection management from the HTable implementation to the end user/client. We previously had very little, if any, hooks to close our HBase tables or attempt to clean up resources. In response to this change, there were a couple options for dealing with this:
1. Completely rewrite our HBase client logic to fully manage connection lifecycle
2. Isolate the connection management change to the existing HTableProvider implementation used throughout Metron and make a smaller, incremental change to set us up for the eventual upgrade to HBase 2.x.

This PR takes the approach in option 2. The biggest question surrounding this approach is whether the included connection management changes introduced in the TableProvider are sufficient, or if we need to immediately take a more robust connection pooling approach to dealing with HBase connections. I spent some time looking at the current HTable implementation that we depend on. Every time an HTable is created, the underlying code makes a call to an internal connection manager. It's unclear to me what the connection management contract is for the user in this case, e.g. stale connection cleanup, connection retries, connection pooling, etc. This is probably the riskiest part of this change. The way that I'm handling this is to 
1. Make the connections thread safe through use of a ThreadLocal connection variable. There are issues with instantiating a ThreadLocal variable by default in code that will be serialized by Storm. ThreadLocal is not serializable. In order to get around a massive API rewrite that would add initialization similar to our other startup hooks, e.g. StellarFunctions.initialize(), I opted for an approach that would allow us to get a similar effect via lazy initialization.
2. Providing some basic connection retry logic that will initiate a new connection if the thread's current connection happens to have closed for whatever reason.

The combination of these 2 options provides a semi-robust way to handle connections without boiling the ocean as well as offering per-thread connection re-use that should limit the overall number of connections we keep open to HBase in a reasonable and reliable way.

I'd like to connect with some HBase committers or just hop on their dev list and ask what the implications of leaving connections alone would be. Again, we don't really do anything to manage creating/closing connections currently, and afaict we have never had any reports of leaking connections. This doesn't mean we won't now, but it's worth noting what the status quo has been for a while now.

**Note: This is a backwards compatible change**

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- n/a If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- n/a Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

- n/a Have you ensured that any documentation diagrams have been updated, along with their source files, using [draw.io](https://www.draw.io/)? See [Metron Development Guidelines](https://cwiki.apache.org/confluence/display/METRON/Development+Guidelines) for instructions.

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-platform/metron-hbase/metron-hbase-common/src/test/java/org/apache/metron/hbase/client/HBaseClientTest.java [hunk 1]
```diff
@@ -86,8 +85,14 @@ public static void startHBase() throws Exception {
...
     util.waitTableEnabled(table.getName());
+    TableProvider provider = new TableProvider() {
+      @Override
+      public Table getTable(Configuration config, String tableName) throws IOException {
+        return table;
+      }
+    };
     // setup the client
-    client = new HBaseClient((c,t) -> table, table.getConfiguration(), tableName);
+    client = new HBaseClient(provider, table.getConfiguration(), tableName);
```
Comment (mmiklavc) [inline metron-platform/metron-hbase/metron-hbase-common/src/test/java/org/apache/metron/hbase/client/HBaseClientTest.java]: I need to revert this change - originally, I add Closeable to the TableProvider interface, which made lambdas unable to work as they currently do. I reverted that change, but haven't reverted the few lambda changes yet. This is partially contingent on how we land on the connection management logic, and whether it's reasonable to not explicitly close those connections.

File: metron-platform/metron-hbase/metron-hbase-common/src/main/java/org/apache/metron/hbase/HTableProvider.java [hunk 2]
```diff
@@ -18,13 +18,49 @@
...
+   */
+  private static Map<Configuration, ThreadLocal<RetryingConnection>> connMap = new ConcurrentHashMap<>();
+
+  @Override
+  public Table getTable(Configuration config, String tableName)
+      throws IOException {
+    return getConnection(config).getTable(TableName.valueOf(tableName));
+  }
+
+  private Connection getConnection(Configuration config) throws IOException {
```
Comment (Apache9) [inline metron-platform/metron-hbase/metron-hbase-common/src/main/java/org/apache/metron/hbase/HTableProvider.java]: Will we use different Configurations here?

File: metron-platform/metron-hbase/metron-hbase-common/src/main/java/org/apache/metron/hbase/HTableProvider.java [hunk 3]
```diff
@@ -18,13 +18,49 @@
...
+      return conn;
     }
+  }
+
+  /**
+   * We have to handle serialization issues with Storm via indirections. Rather than re-implement
+   * the interface everywhere we touch HBase, we can use a lazy initialization scheme to encapsulate
+   * this within the HTableProvider. This is a sort of poor man's connection pool.
+   */
+  private static Map<Configuration, ThreadLocal<RetryingConnection>> connMap = new ConcurrentHashMap<>();
```
Comment (Apache9) [inline metron-platform/metron-hbase/metron-hbase-common/src/main/java/org/apache/metron/hbase/HTableProvider.java]: And actually, Connection is thread safe, so you only need to cache Tables. And since the Table is getting from Connection, which means it will not make new TCP connections to HBaseCluster, maybe we even do not need to cache the Tables, as it is only some in memory operations.

**Comment by mmiklavc** [comment]
Linking this thread from the Apache HBase dev list - https://lists.apache.org/thread.html/6b83cd7548efb8c37899063affc97e4c5ce823a13359a49b477e3c07@%3Cdev.hbase.apache.org%3E

Thanks @Apache9 for the feedback and assistance here - greatly appreciated!

File: metron-platform/metron-hbase/metron-hbase-common/src/main/java/org/apache/metron/hbase/HTableProvider.java (hunk 3 above)
Comment (tigerquoll) [reply metron-platform/metron-hbase/metron-hbase-common/src/main/java/org/apache/metron/hbase/HTableProvider.java]: Plus you are going to leak HBaseConnections here, as they never get cleaned up.

**Comment by mmiklavc** [comment]
> Hbaseconnections are thread-safe objects. Setting up a HBaseConnection is pretty resource intensive. If you are authenticating with a static identity that is not going to change (i.e. a service account), and you are constantly interacting with HBase, then I believe it is recommended that you keep the HBaseConnection object around in a cache or context, and reuse it to pull out Table and Mutation objects.

Yes, that is what we're doing here. The connection model differs subtly from a standard request/response lifecycle web app in that we don't need to open/close the connections. They are bound to the process on startup for the duration of the program's execution. There should be no instance where we are left with a leaked connection because we're only opening a new one when the given thread does not already have an open connection. This generally only happens during application initialization/startup for us.

**Comment by tigerquoll** [comment]
Metron tutorial needs updating from the new Kafka client tools.  Last line of page 1 needs to be updated to:
`${HDP_HOME}/kafka-broker/bin/kafka-console-consumer.sh --bootstrap-server $BROKERLIST --topic squid --from-beginning`

**Comment by tigerquoll** [comment]
Metron tutorial step 5 global.json additional to support squid needs a comma after last square bracket.

**Comment by tigerquoll** [comment]
Metron tutorial Step 5,  no spare workers available to run squid topology.  Need to add instructions on adding another supervisor slot port.

**Comment by tigerquoll** [comment]
Metron tutorial part 2.  To "Check if Zookeeper enrichment tag was properly populated"
command should be changed to
`${METRON_HOME}/bin/zk_load_configs.sh -m DUMP -z $ZOOKEEPER -c ENRICHMENT -n squid`

**Comment by tigerquoll** [comment]
Metron tutorial part 2, should have instructions to retrigger squid data feed, and directions on how to drive the Rest interface to prove data is being enriched.

**Comment by tigerquoll** [comment]
Threatintel tutorial: Dropping of elasticsearch index requires a sudo as their owned by root

**Comment by tigerquoll** [comment]
What commands are you using to query the ES index?

**Comment by mmiklavc** [comment]
I was getting classpath issues with my PR that now seem to be fixed with https://github.com/apache/metron/pull/1498. That PR should get merged tomorrow sometime, barring any discussion to the contrary. In the meantime, I've merged those changes into this PR for sake of testing. I ran through the major parts of the test plan one more time and everything works as expected. I did not repeat the failure recovery scenario.

For reference, this another exception I ran into while running the basic parser test. Again, this issue appears to have been resolved now with the changes in pr 1498:
```
2019-08-27 20:38:55.531 o.a.s.d.executor Thread-12-parserBolt-executor[5 5] [ERROR]
java.lang.IllegalStateException: Unable to process transformation: DOMAIN_REMOVE_SUBDOMAINS(full_hostname) for domain_without_subdomains because Unable to parse: DOMAIN_REMOVE_SUBDOMAINS(full_hostname) due to: EXACT with relevant variables full_hostname=www.aliexpress.com
        at org.apache.metron.common.field.transformation.StellarTransformation.map(StellarTransformation.java:73) ~[stormjar.jar:?]
        at org.apache.metron.common.configuration.FieldTransformer.transform(FieldTransformer.java:139) ~[stormjar.jar:?]
        at org.apache.metron.common.configuration.FieldTransformer.transformAndUpdate(FieldTransformer.java:159) ~[stormjar.jar:?]
        at org.apache.metron.parsers.ParserRunnerImpl.applyFieldTransformations(ParserRunnerImpl.java:299) ~[stormjar.jar:?]
        at org.apache.metron.parsers.ParserRunnerImpl.processMessage(ParserRunnerImpl.java:254) ~[stormjar.jar:?]
        at org.apache.metron.parsers.ParserRunnerImpl.lambda$execute$0(ParserRunnerImpl.java:151) ~[stormjar.jar:?]
        at java.util.ArrayList.forEach(ArrayList.java:1249) ~[?:1.8.0_112]
        at org.apache.metron.parsers.ParserRunnerImpl.execute(ParserRunnerImpl.java:150) ~[stormjar.jar:?]
        at org.apache.metron.parsers.bolt.ParserBolt.execute(ParserBolt.java:257) [stormjar.jar:?]
        at org.apache.storm.daemon.executor$fn__10252$tuple_action_fn__10254.invoke(executor.clj:735) [storm-core-1.1.0.2.6.5.1175-1.jar:1.1.0.2.6.5.1175-1]
        at org.apache.storm.daemon.executor$mk_task_receiver$fn__10171.invoke(executor.clj:466) [storm-core-1.1.0.2.6.5.1175-1.jar:1.1.0.2.6.5.1175-1]
        at org.apache.storm.disruptor$clojure_handler$reify__9685.onEvent(disruptor.clj:40) [storm-core-1.1.0.2.6.5.1175-1.jar:1.1.0.2.6.5.1175-1]
        at org.apache.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:472) [storm-core-1.1.0.2.6.5.1175-1.jar:1.1.0.2.6.5.1175-1]
        at org.apache.storm.utils.DisruptorQueue.consumeBatchWhenAvailable(DisruptorQueue.java:451) [storm-core-1.1.0.2.6.5.1175-1.jar:1.1.0.2.6.5.1175-1]
        at org.apache.storm.disruptor$consume_batch_when_available.invoke(disruptor.clj:73) [storm-core-1.1.0.2.6.5.1175-1.jar:1.1.0.2.6.5.1175-1]
        at org.apache.storm.daemon.executor$fn__10252$fn__10265$fn__10320.invoke(executor.clj:855) [storm-core-1.1.0.2.6.5.1175-1.jar:1.1.0.2.6.5.1175-1]
        at org.apache.storm.util$async_loop$fn__553.invoke(util.clj:484) [storm-core-1.1.0.2.6.5.1175-1.jar:1.1.0.2.6.5.1175-1]
        at clojure.lang.AFn.run(AFn.java:22) [clojure-1.7.0.jar:?]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_112]
Caused by: org.apache.metron.stellar.dsl.ParseException: Unable to parse: DOMAIN_REMOVE_SUBDOMAINS(full_hostname) due to: EXACT with relevant variables full_hostname=www.aliexpress.com
        at org.apache.metron.stellar.common.BaseStellarProcessor.createException(BaseStellarProcessor.java:173) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.common.BaseStellarProcessor.parse(BaseStellarProcessor.java:154) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.common.CachingStellarProcessor.parseUncached(CachingStellarProcessor.java:161) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.common.CachingStellarProcessor.parse(CachingStellarProcessor.java:155) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.common.field.transformation.StellarTransformation.map(StellarTransformation.java:50) ~[stormjar.jar:?]
        ... 18 more
Caused by: java.lang.NoSuchFieldError: EXACT
        at org.apache.metron.guava.17.0.net.InternetDomainName.findPublicSuffix(InternetDomainName.java:173) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.guava.17.0.net.InternetDomainName.<init>(InternetDomainName.java:158) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.guava.17.0.net.InternetDomainName.from(InternetDomainName.java:213) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.dsl.functions.NetworkFunctions.toDomainName(NetworkFunctions.java:271) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.dsl.functions.NetworkFunctions.access$000(NetworkFunctions.java:34) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.dsl.functions.NetworkFunctions$RemoveSubdomains.apply(NetworkFunctions.java:86) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.dsl.BaseStellarFunction.apply(BaseStellarFunction.java:30) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.common.StellarCompiler.lambda$exitTransformationFunc$13(StellarCompiler.java:664) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.common.StellarCompiler$Expression.apply(StellarCompiler.java:259) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.common.BaseStellarProcessor.parse(BaseStellarProcessor.java:151) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.common.CachingStellarProcessor.parseUncached(CachingStellarProcessor.java:161) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.stellar.common.CachingStellarProcessor.parse(CachingStellarProcessor.java:155) ~[dep-stellar-common-0.7.2-uber-6958fda1-1cb6-4e03-80ba-c5171221d03a.jar.1566935872000:?]
        at org.apache.metron.common.field.transformation.StellarTransformation.map(StellarTransformation.java:50) ~[stormjar.jar:?]
        ... 18 more
```

**Comment by tigerquoll** [comment]
For the test script: full dev still needed another supervisor slot in Storm to allow the squid enrichment topology to run.

**Comment by mmiklavc** [comment]
@tigerquoll what all are you running? This leaves me with 1 slot available still
image

**Comment by mmiklavc** [comment]
@tigerquoll I've updated the test script on this PR to reflect the note in the first tutorial.

**Comment by MohanDV** [comment]
@mmiklavc I have completed tests on a 5 node cluster, for the profiler tests I see below exceptions in the worker log of enrichment 
`java.lang.IllegalStateException: ENRICHMENT error with stellar failed: Unable to parse: PROFILE_GET( 'hello-world', ip_dst_addr, PROFILE_FIXED(2, 'MINUTES')) due to: null with relevant variables ip_dst_addr=missing
	at org.apache.metron.enrichment.parallel.ParallelEnricher.lambda$apply$0(ParallelEnricher.java:195) ~[stormjar.jar:?]
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590) ~[?:1.8.0_112]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_112]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_112]`

**Comment by mmiklavc** [comment]
@MohanDV thanks for the review! The enrichment I constructed wasn't robust in the sense that if an ip_dst_addr is missing, it would be expected to blow up like this. The data samples I provided for squid should all have had an ip_dst_addr. This looks like one of the records may be null for some reason or another. Can you share the sample squid access log? I suspect there's a record in there that doesn't look right.

**Comment by MohanDV** [comment]
@mmiklavc Yes I have an entry `1567166198.507      0 ::1 TAG_NONE/400 3983 GET ;i=1;doned42.7630722pwdd-88.2142563 - HIER_NONE/- text/html
` in the squid access log for 
`squidclient "https://www.google.com/maps/place/Waterford,+WI/@42.7639877,-88.2867248,12z/data=!4m5!3m4!1s0x88059e67de9a3861:0x2d24f51aad34c80b!8m2!3d42.7630722!4d-88.2142563"
`

**Comment by mmiklavc** [comment]
@MohanDV that makes sense then - this would be an expected exception in that case.

**Comment by tigerquoll** [comment]
Tests completed on a fresh pull / full dev deployment.  All tests passed successfully.  One nit:  Command for stopping unneeded parsers should be: 
`for parser in bro__snort__yaf profiler pcap batch_indexing; do storm kill $parser; done`

**Comment by tigerquoll** [comment]
One thing to note, there was a non-fatal class cast exception in the nimbus logs:
```
2019-09-02 01:51:04.756 o.a.s.m.ClusterMetricsConsumerExecutor main [ERROR] Could not instantiate or prepare Cluster Metrics Consumer with fully qualified name org.apache.storm.metric.LoggingMetricsConsumer
java.lang.ClassCastException: org.apache.storm.metric.LoggingMetricsConsumer cannot be cast to org.apache.storm.metric.api.IClusterMetricsConsumer
```

**Comment by MohanDV** [comment]
Thanks @mmiklavc ! it's +1 with verification from my side .
